### PR TITLE
PMM-1091: --json and --format support for `pmm-admin list`

### PR DESCRIPTION
**Pull request by arvenil**
Two new options for `pmm-admin list`:
```
--json            print result as json
```
and one which [might be familiar to you](https://docs.docker.com/engine/admin/formatting/) if you are using `docker` cli
```
--format string   print result using a Go template
```

```console
$ pmm-admin list --help
This command displays the list of monitoring services and their details.

Usage:
  pmm-admin list [flags]

Aliases:
  list, ls

Flags:
      --format string   print result using a Go template
  -h, --help            help for list
      --json            print result as json

Global Flags:
  -c, --config-file string   PMM config file (default "/tmp/pmm-client-test-rootdir-995307892/usr/local/percona/pmm-client/pmm.yml")
      --verbose              verbose output
```

```console
$ pmm-admin list
pmm-admin gotest

PMM Server      | 127.0.0.1:53629 
Client Name     | test-client-name
Client Address  | empty (data)
Service Manager | darwin-launchd

---------------- ----------------- ----------- -------- ------------ --------
SERVICE TYPE     NAME              LOCAL PORT  RUNNING  DATA SOURCE  OPTIONS 
---------------- ----------------- ----------- -------- ------------ --------
mongodb:queries  test-client-name  -           YES                 -                    
mysql:queries    test-client-name  -           YES                 -                    
```

```console
$ pmm-admin list --json  # pmm-admin list --format '{{json .}}'
{"Version":"gotest","Server Address":"127.0.0.1:53629","Server Security":"","Client Name":"test-client-name","Client Address":"empty","Client Bind Address":"(data)","Platform":"darwin-launchd","Err":"","Services":[{"Type":"mongodb:queries","Name":"test-client-name","Port":"-","Running":true,"DSN":"-","Options":"","SSL":"","Password":""},{"Type":"mysql:queries","Name":"test-client-name","Port":"-","Running":true,"DSN":"-","Options":"","SSL":"","Password":""}]}
```

```console
$ pmm-admin list --format '{{json .}}' | jq
{
  "Version": "gotest",
  "ServerAddress": "127.0.0.1:53629",
  "ServerSecurity": "",
  "ClientName": "test-client-name",
  "ClientAddress": "empty",
  "ClientBindAddress": "(data)",
  "Platform": "darwin-launchd",
  "Err": "",
  "Services": [
    {
      "Type": "mongodb:queries",
      "Name": "test-client-name",
      "Port": "-",
      "Running": true,
      "DSN": "-",
      "Options": "",
      "SSL": "",
      "Password": ""
    },
    {
      "Type": "mysql:queries",
      "Name": "test-client-name",
      "Port": "-",
      "Running": true,
      "DSN": "-",
      "Options": "",
      "SSL": "",
      "Password": ""
    }
  ]
}
```

With `--format` you can pick what you want
```console
$ pmm-admin list --format '{{.ClientName}}'
test-client-name
```

```console
$ pmm-admin list --format 'SERVICE TYPE NAME    LOCAL PORT      RUNNING DATA SOURCE     OPTIONS
{{range .Services}}{{.Type}}    {{.Name}}       {{.Port}}       {{if .Running}}YES{{else}}NO{{end}}     {{.DSN}}        {{.Options}}
{{end}}'
SERVICE TYPE           NAME                    LOCAL PORT        RUNNING        DATA SOURCE        OPTIONS
mongodb:queries        test-client-name        -                 YES            -                  
mysql:queries          test-client-name        -                 YES            -                  
```

```console
$ pmm-admin list --format '{{json .Services}}'
[{"Type":"mongodb:queries","Name":"test-client-name","Port":"-","Running":true,"DSN":"-","Options":"","SSL":"","Password":""},{"Type":"mysql:queries","Name":"test-client-name","Port":"-","Running":true,"DSN":"-","Options":"","SSL":"","Password":""}]
```

```console
$ pmm-admin list --format '{{json .Services}}' | jq
[
  {
    "Type": "mongodb:queries",
    "Name": "test-client-name",
    "Port": "-",
    "Running": true,
    "DSN": "-",
    "Options": "",
    "SSL": "",
    "Password": ""
  },
  {
    "Type": "mysql:queries",
    "Name": "test-client-name",
    "Port": "-",
    "Running": true,
    "DSN": "-",
    "Options": "",
    "SSL": "",
    "Password": ""
  }
]
```

### CODE REVIEW

**Comment by AlekSi** [comment]
Really concerned about a huge number of new and update dependencies. Do we need them? Do we need `--format` at all?

**Comment by arvenil** [comment]
> Do we need them? Do we need --format at all?

Why we don't need `--format`? It's for free if you use https://golang.org/pkg/text/template/ and you can get any data you wish with it. It's also familiar to people using docker cli (maybe there are other cli tools using that too?).

> Really concerned about a huge number of new and update dependencies.

`+193,824 −660,402`
* There is only one new tiny dependency https://github.com/docker/cli/blob/master/templates/templates.go
* I've removed unused pkgs that's why 3 times more removals - can push that to separate PR if this bothers you.
* Not sure why we use `--strip-vendor`, it brakes dependencies, was forced to manually lock `github.com/hashicorp/raft` version which is a dependency of dependency and was vendored in sub dir.
* Some dependencies got updates, with vendor dir commit to source tree this is visible "noise". What you suggest? Don't update deps? Get rid of vendor dir? ;)

**Comment by arvenil** [comment]
@AlekSi 
> Do we need them? Do we need --format at all?

Why we don't need `--format`? It's for free if you use https://golang.org/pkg/text/template/ and you can get any data you wish with it. It's also familiar to people using docker cli (maybe there are other cli tools using that too?).

> Really concerned about a huge number of new and update dependencies.

`+193,824 −660,402`
* There is only one new tiny dependency https://github.com/docker/cli/blob/master/templates/templates.go
* I've removed unused pkgs that's why 3 times more removals - can push that to separate PR if this bothers you.
* Not sure why we use `--strip-vendor`, it brakes dependencies, was forced to manually lock `github.com/hashicorp/raft` version which is a dependency of dependency and was vendored in sub dir.
* Some dependencies got updates, with vendor dir commit to source tree this is visible "noise". What you suggest? Don't update deps? Get rid of vendor dir? ;)

File: pmm/list.go [hunk 1]
```diff
@@ -50,32 +53,143 @@ func (s sortOutput) Swap(i, j int) {
 }
 
 func (s sortOutput) Less(i, j int) bool {
-	if strings.Compare(s[i].Port, s[j].Port) == -1 {
-		return true
+	if s[i].Port != s[j].Port {
```
Comment (percona-csalguero) [inline pmm/list.go]: Just a suggestion for future releases:
Sort by Type, then Name and then Port

**Comment by AlekSi** [comment]
> Why we don't need --format?

No-one asked for it.

> It's for free if you use https://golang.org/pkg/text/template/

Again, no-one asked for templating. `json.Marshal()` is all we need there. Let's not add features just because they are cool.

> I've removed unused pkgs that's why 3 times more removals - can push that to separate PR if this bothers you.

Those deps were added for gRPC/pmm-managed. I made a mistake adding them to `master` without usage. I will revert that commit.

> Not sure why we use --strip-vendor

https://github.com/mattfarina/golang-broken-vendor. It caused problems previously.

> Some dependencies got updates, with vendor dir commit to source tree this is visible "noise". What you suggest? Don't update deps? Get rid of vendor dir? ;)

I suggest not to update other deps when you add a new one.

In summary, we can slate that PR until or 1.4 – our you can implement just `--json`, without new deps, for 1.3.

**Comment by arvenil** [comment]
It's not cool, can be useful for services, to avoid parsing json you can simply ask for value you want.

I can remove `--format`if you wish, doesn't matter as internally it's all going to be the same.

**Comment by arvenil** [comment]
> Again, no-one asked for templating. json.Marshal() is all we need there. Let's not add features just because they are cool.

Did you looked at original code? On what you want to call json.Marshal()?

**Comment by AlekSi** [comment]
> It's not cool, can be useful for services

I just want to say that they did not asked for that. It may or may not be useful for them – we don't know. And if it doesn't, then we introduced both internal (a lot of code) and external (new flag) complexity for nothing.

> I can remove --formatif you wish, doesn't matter as internally it's all going to be the same. --format is basically side effect of adding json.

That will keep internal complexity.

> Did you looked at original code? On what you want to call json.Marshal()?

That's exatly what I thought you will change.
My idea was to change list method (and eventualy other methods) to return data objects and errors, and do not log anything. Message output could be moved to commands – and there we can make any output we want. So I was thinking about separating logic and representation. We can do that for list only in 1.3 (only JSON, without deps), or discuss it and implement for everything in 1.4.

@rnovikovP WDYT?

**Comment by arvenil** [comment]
@aleksi

> I can remove --formatif you wish, doesn't matter as internally it's all going to be the same. --format is basically side effect of adding json.
> 
> That will keep internal complexity.

What internal complexity you are referring to? Are you referring to standard golang templates?

>     Did you looked at original code? On what you want to call json.Marshal()?
> 
> That's exatly what I thought you will change.
> My idea was to change list method (and eventualy other methods) to return data objects and errors, and do not log anything. Message output could be moved to commands – and there we can make any output we want. So I was thinking about separating logic and representation. We can do that for list only in 1.3 (only JSON, without deps), or discuss it and implement for everything in 1.4.

And exactly this is done. Instead of printing to stdout, I'm creating object, and then you can format the object e.g. with a template. You could also json.Marshal() that object, but since we already do templating you can just use json as directive, it's 5 lines of code and you can print any data as json.

From technical point this PR:
* Create data object instead of printing to stdout
* Since we have now objects we can't just print them, but we can use golang standard templates (percona-toolkit is using them already, same as docker) and tabwriter (used by docker). So instead of
```
	// Print table.
	maxTypeLen := len("SERVICE TYPE")
	maxNameLen := len("NAME")
	maxDSNlen := len("DATA SOURCE")
	maxOptsLen := len("OPTIONS")
	for _, in := range svcTable {
		if len(in.Type) > maxTypeLen {
			maxTypeLen = len(in.Type)
		}
		if len(in.Name) > maxNameLen {
			maxNameLen = len(in.Name)
		}
		if len(in.DSN) > maxDSNlen {
			maxDSNlen = len(in.DSN)
		}
		if len(in.Options) > maxOptsLen {
			maxOptsLen = len(in.Options)
		}
	}
	maxTypeLen++
	maxNameLen++
	maxDSNlen++
	maxOptsLen++
	maxStatusLen := 8

	fmtPattern := "%%-%ds %%-%ds %%-%ds %%-%ds %%-%ds %%-%ds\n"
	linefmt := fmt.Sprintf(fmtPattern, maxTypeLen, maxNameLen, 11, maxStatusLen, maxDSNlen, maxOptsLen)

	fmt.Printf(linefmt, strings.Repeat("-", maxTypeLen), strings.Repeat("-", maxNameLen), strings.Repeat("-", 11),
		strings.Repeat("-", maxStatusLen), strings.Repeat("-", maxDSNlen), strings.Repeat("-", maxOptsLen))
	fmt.Printf(linefmt, "SERVICE TYPE", "NAME", "LOCAL PORT", "RUNNING", "DATA SOURCE", "OPTIONS")
	fmt.Printf(linefmt, strings.Repeat("-", maxTypeLen), strings.Repeat("-", maxNameLen), strings.Repeat("-", 11),
		strings.Repeat("-", maxStatusLen), strings.Repeat("-", maxDSNlen), strings.Repeat("-", maxOptsLen))

	sort.Sort(sortOutput(svcTable))
	maxStatusLen += 11
	linefmt = fmt.Sprintf(fmtPattern, maxTypeLen, maxNameLen, 11, maxStatusLen, maxDSNlen, maxOptsLen)
	for _, i := range svcTable {
		fmt.Printf(linefmt, i.Type, i.Name, i.Port, colorStatus("YES", "NO", i.Status), i.DSN, i.Options)
	}
	fmt.Println()

```
you can have template
```
ERVICE TYPE NAME    LOCAL PORT      RUNNING DATA SOURCE     OPTIONS
{{range .Services}}{{.Type}}    {{.Name}}       {{.Port}}       {{if .Running}}YES{{else}}NO{{end}}     {{.DSN}}        {{.Options}}
{{end}}
```
* For json, objects are just json.Marshaled()
* Since we have templates, with few lines of code you can generate json that suits your need for example if you need just the list of Services without status header
```
$ pmm-admin list --format '{{json .Services}}' | jq
[
  {
    "Type": "mongodb:queries",
    "Name": "test-client-name",
    "Port": "-",
    "Running": true,
    "DSN": "-",
    "Options": "",
    "SSL": "",
    "Password": ""
  },
  {
    "Type": "mysql:queries",
    "Name": "test-client-name",
    "Port": "-",
    "Running": true,
    "DSN": "-",
    "Options": "",
    "SSL": "",
    "Password": ""
  }
]
```

**Comment by AlekSi** [comment]
Could merge #83, rebase that PR on top of it and make sure other deps are not changed?

**Comment by arvenil** [comment]
> Could merge #83, rebase that PR on top of it and make sure other deps are not changed?

As I said, if this bothers you I will prepare separate PRs, #83 doesn't remove everything. I'm not going to loose time on merging this.
